rfcs/20210513-pluggable-profiler-for-tensorflow.md

```diff
@@ -0,0 +1,429 @@
+         				  &size_in_bytes, c_status.get());
+        // Deserialize XSpace from the buffer and return it.
+        XSpace plugin_space;
+        plugin_space.ParseFromArray(buffer.data(), buffer.size());
+        for (XPlane& plugin_plane: *plugin_space.mutable_planes()) {
+          XPlane* plane = space->add_planes();
+          plane->Swap(&plugin_plane);
+        }
+      }
+      return tensorflow::StatusFromTF_Status(c_status.get());
+    }
+  
+   private:
+    PluggableProfiler(TP_ProfilerFns* profiler_fns, TP_Profiler* profiler) 
+      : profiler_fns_(profiler_fns), profiler_(profiler) {}
+    
+    TP_ProfilerFns* profiler_fns_;
+    TP_Profiler* profiler_;
+  }
+  ```
+
+* **PluggableProfiler Initialization and Registration**
+
+  Core TensorFlow will load `TF_InitProfiler` from plugin's dynamic library installed under "…python_dir.../site-packages/tensorflow-plugins" and pass the address of `TF_InitProfiler` symbol to `InitPluginProfiler` to do initialization and registration. TensorFlow retrieves `TF_ProfilerRegistrationParams` from the plugin and does the compatibility checks. If passed, TensorFlow will register the PluggableProfiler creation function to the [ProfilerFactroy](https://github.com/tensorflow/tensorflow/blob/aa855d21ac89e6649ec782ef4efd4e126b37d79d/tensorflow/core/profiler/lib/profiler_factory.cc#L39).
+
+  ```c++
+  class PluggableProfilerFactory{
+   public:
+    PluggableProfilerFactory(TP_Profiler profiler,
+  			   void (*destroy_profiler)(TP_Profiler*),
+  			   TP_ProfilerFns profiler_fns,
+  			   void (*destroy_profiler_fns)(TP_ProfilerFns*))
+    	: profiler_(std::move(profiler)),
+            destroy_profiler_(destroy_profiler),
+            profiler_fns_(std::move(profiler_fns)),
+            destroy_profiler_fns_(destroy_profiler_fns) {}
+  
+    ~PluggableProfilerFactory() {
+      destroy_profiler_(&profiler_);
+      destroy_profiler_fns_(&profiler_fns_);
+
+    std::unique_ptr<tensorflow::profiler::ProfilerInterface> CreatePluggableProfiler(
+        const ProfileOptions& options) {
+      return PluggableProfiler::CreatePluggableProfiler(options, &profiler_,
+                                                      &profiler_fns_);
+    }
+
+   private:
+      TP_Profiler profiler_{TP_PROFILER_STRUCT_SIZE};
+      void (*destroy_profiler_)(TP_Profiler*);
+      TP_ProfilerFns profiler_fns_{TP_PROFILER_FNS_STRUCT_SIZE};
+      void (*destroy_profiler_fns_)(TP_ProfilerFns*);
+    }
+  
+  }
+  
+  Status InitPluginProfiler(TFInitProfilerFn init_fn) {
+    TF_ProfilerRegistrationParams params{TF_PROFILER_REGISTRATION_PARAMS_STRUCT_SIZE};
+    TP_Profiler profiler{TP_PROFILER_STRUCT_SIZE};
+    TP_ProfilerFns profiler_fns{TP_PROFILER_FNS_STRUCT_SIZE};
+    params.major_version = TP_MAJOR;
+    params.minor_version = TP_MINOR;
+    params.patch_version = TP_PATCH;
+    params.profiler = &profiler;
+    params.profiler_fns = &profiler_fns;
+    OwnedTFStatus c_status(TF_NewStatus());
```

Comment:
std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());

for consistency with the rest of the code

Reply:
Done